pallets/collective-proxy/src/lib.rs

```diff
@@ -121,5 +179,68 @@ pub mod pallet {
 
             Ok(())
         }
+
+        /// Register a proxy account for the sender that is able to make calls on its behalf.
+        ///
+        /// The dispatch origin for this call must be _Signed_.
+        ///
+        /// Parameters:
+        /// - `proxy`: The account that the `caller` would like to make a proxy.
+        /// - `filter`: Call filter used for the proxy
+        #[pallet::call_index(1)]
+        #[pallet::weight(T::WeightInfo::add_proxy(T::MaxProxies::get()))]
+        pub fn add_proxy(
+            origin: OriginFor<T>,
+            proxy: T::AccountId,
+            filter: T::CallFilter,
+        ) -> DispatchResult {
+            T::ProxyAdmin::ensure_origin(origin)?;
+            Proxies::<T>::try_mutate(|proxies| -> Result<(), DispatchError> {
+                if !proxies.iter().any(|p| p.proxy == proxy && p.filter.is_superset(&filter)) {
```

Comment:
The default implementation of `is_superset` returns **false** if no custom logic is provided - [here](https://github.com/paritytech/polkadot-sdk/blob/e79fd2bb9926be7d94be0c002676fca57b6116bf/substrate/frame/support/src/traits/filter.rs#L85C5-L85C16).

Imagine, if `Proxies` contains:

```rs
Proxies = [
    ProxyDefinition { proxy: Account1, filter: Filter1 }
]
```

and you add **Filter2**, which is a superset of **Filter1**, the `add_proxy` logic will push it into `Proxies`:

```rs
Proxies = [
    ProxyDefinition { proxy: Account1, filter: Filter1 },
    ProxyDefinition { proxy: Account1, filter: Filter2 }
]
```

However, the `find_proxy` will still return the outdated proxy:

```rs
ProxyDefinition { proxy: Account1, filter: Filter1 }
```

Maybe this could lead to incorrect permissions being applied. Adding a test case to verify and address this scenario would be highly beneficial.

Reply:
Hey Igor! I reworked storage to map to address your (and Dino's) comments.